syntax = "proto3";

import "google/protobuf/descriptor.proto";
import "google/protobuf/empty.proto";

package service_kit;

message ServiceIdentify {
  string name = 1;
}

message ServiceDefinition {
  string name = 1;
  int32 port = 2;
}

message ServiceDefinitionWithSignature {
  string name = 1;
  int32 port = 2;
  google.protobuf.ServiceDescriptorProto proto = 3;
}

message RegisterResult {
  bool ok = 1;
  string message = 2;
}

service ServiceRegistry {
  rpc Register (ServiceDefinitionWithSignature) returns (RegisterResult) {}
  rpc HeartBeat (ServiceDefinition) returns (google.protobuf.Empty) {}
  rpc Unregister (ServiceDefinition) returns (google.protobuf.Empty) {}
}

message ServiceProvider {
  string address = 1;
}

service ServiceManager {
  rpc Query(ServiceIdentify) returns (ServiceProvider) {}
}
